spatial/r3/box_test.go

```diff
@@ -0,0 +1,200 @@
+// license that can be found in the LICENSE file.
+
+package r3
+
+import (
+	"testing"
+
+	"golang.org/x/exp/rand"
+)
+
+func TestBoxContains(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < 200; i++ {
+		b := randomBox(rnd)
+		for j := 0; j < 10; j++ {
+			contained := b.random(rnd)
+			if !b.Contains(contained) {
+				t.Error("bounding box should contain Vec")
+			}
+		}
+		uncontained := [6]Vec{
+			Add(b.Max, Vec{1, 0, 0}),
+			Add(b.Max, Vec{0, 1, 0}),
+			Add(b.Max, Vec{0, 0, 1}),
+			Sub(b.Min, Vec{1, 0, 0}),
+			Sub(b.Min, Vec{0, 1, 0}),
+			Sub(b.Min, Vec{0, 0, 1}),
+		}
+		for _, unc := range uncontained {
+			if b.Contains(unc) {
+				t.Error("box should not contain vec")
+			}
+		}
+	}
+}
+
+func TestBoxUnion(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < 200; i++ {
+		b1 := randomBox(rnd)
+		b2 := randomBox(rnd)
+		u := b1.Union(b2)
+		for j := 0; j < 10; j++ {
+			contained := b1.random(rnd)
+			if !u.Contains(contained) {
+				t.Error("union should contain b1's Vec")
+			}
+			contained = b2.random(rnd)
+			if !u.Contains(contained) {
+				t.Error("union should contain b2's Vec")
+			}
+		}
+		uncontained := [6]Vec{
+			Add(maxElem(b1.Max, b2.Max), Vec{1, 0, 0}),
+			Add(maxElem(b1.Max, b2.Max), Vec{0, 1, 0}),
+			Add(maxElem(b1.Max, b2.Max), Vec{0, 0, 1}),
+			Sub(minElem(b1.Min, b2.Min), Vec{1, 0, 0}),
+			Sub(minElem(b1.Min, b2.Min), Vec{0, 1, 0}),
+			Sub(minElem(b1.Min, b2.Min), Vec{0, 0, 1}),
+		}
+		for _, unc := range uncontained {
+			if !b1.Contains(unc) && !b2.Contains(unc) && u.Contains(unc) {
+				t.Error("union should not contain Vec")
+			}
+		}
+	}
+}
+
+func TestBoxCenter(t *testing.T) {
+	const tol = 1e-11
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < 300; i++ {
+		b := randomBox(rnd)
+		center := b.Center()
+		size := b.Size()
+		newBox := centeredBox(center, size)
+		if !vecApproxEqual(b.Min, newBox.Min, tol) {
+			t.Errorf("min values of box not equal. got %g, expected %g", newBox.Min, b.Min)
+		}
+		if !vecApproxEqual(b.Max, newBox.Max, tol) {
+			t.Errorf("max values of box not equal. got %g, expected %g", newBox.Max, b.Max)
+		}
+	}
+}
+
+func TestBoxScale(t *testing.T) {
+	const tol = 1e-11
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < 300; i++ {
+		b := randomBox(rnd)
+		size := b.Size()
+
+		scaler := absElem(randomVec(rnd))
+		scaled := b.Scale(scaler)
+		gotScaler := divElem(scaled.Size(), size)
+		if !vecApproxEqual(scaler, gotScaler, tol) {
+			t.Errorf("got scaled %g, expected %g", gotScaler, scaler)
+		}
+		center := b.Center()
+		scaledCenter := scaled.Center()
+		if !vecApproxEqual(center, scaledCenter, tol) {
+			t.Error("scale modified center")
+		}
+	}
+}
+
+func TestBoxVertices(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < 300; i++ {
+		b := randomBox(rnd)
+		gots := b.Vertices()
+		wants := goldenVertices(b)
+		if len(gots) != len(wants) {
+			t.Fatal("bad length of vertices. expect 8, got ", len(gots))
```

Comment:
The addition of a space to make a `Fatal` call format correctly is a smell. I'd prefer `Fatalf` here.